drivers/rpmsg/Kconfig

```diff
@@ -57,6 +57,15 @@ config RPMSG_SYSLOG
 	  We can even cat /sys/kernel/debug/remoteproc/remoteprocX/trace0 to
 	  get the last log which doesn't output yet.
 
+config RPMSG_USRSOCK
+	tristate "RPMSG socket API redirection"
+	depends on RPMSG
+	depends on NET
+	help
+	  Say Y here to redirect sock API from the remote processor.
+	  This driver could give the remote processor the capability
+	  to access the internet through any host network interface.
```

Comment:
I just don't understand this explanation either.

Reply:
Normally, the small DSP/MCU in SoC can't directly access Ethernet/WiFi/Modem hardware and difficult to integrate a full TCP/IP stack.
But, the big core should have some method to access the Internet in general.
With this driver, DSP/MCU could write a simple proxy layer which forward all socket API to the big core and utilize the big core's TCP/IP stack and hardware.
RTOS plus rpmsg-hostfs and rpmsg-usrsock give DSP/MCU a full operating system environment much like Linux.

Reply:
I understand this example.
I get how this rpmsg mechanism might be used to request services from the kernel, e.g. the clocks - with the caveat that this is far from trivial and might have security impacts.

However I don't understand how this might be used for audio stuff typically handled by the SOF firmware (pipelines, processing, etc). On most platforms I am aware of, the host runs Linux and there is a small DSP/MCU taking care of audio tasks. That latter unit is under complete control of the host and will run some sort of RTOS/scheduler along with the SOF firmware.

In short, I don't see what SOF and this rpmsg IPC have to do with each other.
Not trying to be painful, just completely missing the concept.

Reply:
Yes, rpmsg-usrsock isn't directly relate to audio stuff, but other driver could be used in audio:
1.rpmsg-syslog could replace sof\src\include\sof\trace.h, the benifit is that:
  a.We can log the full message instead three chars
  b.We can directly see the log as normal without rmbox
  c.The implementation is more easy to port to other platform
2.With rpmsg-tty, any terminal tool from host side could connect the RTOS shell running on MCU/DSP. this is very useful for developer to debug the problem.
3.rpmsg-hostfs make MCU/DSP access the host file system, for audio:
  a.Load the tuning parameter from file.
  b.Save the tuning parameter to file during manufacturing.
  c.Dump audio pcm to file for debugging and tuning.
  d.Load audio file from host for testing
  e.Dynamiclly load module(.so) from host.
4.rpmsg-clk could control the clock on the host side.
And this is just the first patchset, in the next two patchset, we will show:
1.How to load the firmware from Linux kernel, this part normally platform specific, but which is totally decouple from rpmsg driver layer.
2.Our topology implemenation base on rpmsg framework which show that we can just focus on audio part, and decouple from the firmware loader and common service.
Since SOF is target for all vendor, it's important to ensure that audio stuff as platform independent as possible from both the kernel side and RTOS side.
With remoteproc/rpmsg framework, each vendor just need implement the remoteproc interface in drivers/remoteproc, all other part could be reused without any modification.
The flexible framework also make it's very easy to support the combined use case which is also important since one MCU/DSP may take several role(e.g. audio plus sensor hub).